text/0000-pack-buildpack-new-alternatives.md

```diff
@@ -0,0 +1,308 @@
+
+We will implement the project scaffolding logic in a Go module separate from `pack`.  This decouples `pack` from the project scaffolding implementation.
+
+# Definitions
+[definitions]: #definitions
+
+* **project scaffolding**: minimal code and file system structure used to start a project.
+
+# Motivation
+[motivation]: #motivation
+
+The creation of buildpacks in languages other than `bash` is undocumented in the [Buildpack Author Guide](https://buildpacks.io/docs/buildpack-author-guide).  In particular, creating a buildpack in Go using `libcnb` is undocumented because there is no mechanism to generate a scaffolded project.  This proposal adds a mechanism to `pack` that would allow the buildpacks project to provide `libcnb` buildpack project scaffolding.
+
+`pack buildpack create` supports end-users who wish to scaffold a new buildpack.  The operation of `pack buildpack create` requires Internet access to succeed.  `pack buildpack create --template URL` allows projects (such as Paketo, Heroku and others) to define skeleton projects for new buildpacks.
+
+Replacing `pack buildpack new` with `pack buildpack create` allows buildpack authors to more-easily create new buildpacks in their chosen language and framework.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Project scaffolding is [very](https://cookiecutter.readthedocs.io/) [popular](https://yeoman.io/) [in](https://github.com/kowainik/summoner) [other](https://crates.io/crates/cargo-scaffold) [ecosystems](https://github.com/golang-standards/project-layout).  Scaffolding systems are employed to ease onboarding of new developers.  Within `pack` this feature is targeted at onboarding buildpack authors.
+
+`pack buildpack create` creates the same buildpack scaffolding as `pack buildpack new`.  However, `pack buildpack create --template URL` can be used to access alternative project scaffolding.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The design is modeled on [cookiecutter](https://cookiecutter.readthedocs.io/en/1.7.2/) with reference to [springerle](https://github.com/carlmjohnson/springerle) -- a similar implementation in golang -- and [create-go-app](https://github.com/create-go-app/).  We do not want to `os.Exec` a python subprocess to run cookiecutter as this would require availability of a python runtime environment.  Instead we propose to borrow heavily from create-go-app, and generate the default shell scaffolding from a cloned git repoistory.
+
+The `pack buildpack new` project scaffolding accepts `--api`, `--path` and `--stacks` command line flags.  These command line flags will be preserved in `pack buildpack create`.  However, where a flag is omitted, the end-user will be prompted to provide an appropriate value.
```

Comment:
I don't think we should preserve the `--api` or `--stacks` flags. The purpose of these flags is to provide data to templates, right? So in my mind, that conflicts with Overrides which are also to override values from templates.

It also provides a weird experience, IMHO, in that we collect these values by default but do not provide these values automatically to the template. The template author still needs to add an entry to prompts.toml for them (at least that's how I'm reading it, please correct me if I'm misunderstanding that).

I favor just dropping these arguments, but I could see an alternative approach being that we automatically provide those two values to every template and the template author doesn't need to include them in prompts.toml.

Reply:
We've now dropped these CLI flags.